plugins/mediatek-scaler/fu-mediatek-scaler-device.c

```diff
@@ -0,0 +1,1063 @@
+				fu_device_sleep(FU_DEVICE(self), 1);
+			}
+
+			/* update progress */
+			fu_progress_set_percentage_full(fu_progress_get_child(progress),
+							(gsize)i + 1,
+							(gsize)chunks_sz);
+		}
+
+		/* exit the try loop when successes */
+		fu_device_sleep(FU_DEVICE(self), FU_MEDIATEK_SCALER_DDC_MSG_DELAY_MS);
+		if (fu_mediatek_scaler_device_check_sent_info(self, fw, &error_local))
+			return TRUE;
+
+		if (!g_error_matches(error_local, FWUPD_ERROR, FWUPD_ERROR_WRITE)) {
+			g_propagate_error(error, g_steal_pointer(&error_local));
+			return FALSE;
+		}
+
+		g_debug("retry write_firmware: step: %d, max: %d", retry, DDC_RW_MAX_RETRY_CNT);
+	}
+	g_set_error_literal(error, FWUPD_ERROR, FWUPD_ERROR_INTERNAL, "maximum tries exceeded");
+	return FALSE;
+}
+
+static gboolean
+fu_mediatek_scaler_device_attach(FuDevice *device, FuProgress *progress, GError **error)
+{
+	FuMediatekScalerDevice *self = FU_MEDIATEK_SCALER_DEVICE(device);
+	g_autoptr(FuDeviceLocker) locker = NULL;
+	guint max_tries = 30;
+
+	/* open device */
+	locker = fu_device_locker_new(self->i2c_dev, error);
+	if (locker == NULL)
+		return FALSE;
+
+	/* set the target address */
+	if (!fu_mediatek_scaler_ensure_device_address(self, DDC_DISPLAY_I2C_ADDR >> 1, error))
+		return FALSE;
+
+	/* reboot the device */
+	if (!(fu_mediatek_scaler_device_set_isp_reboot(self, error)))
+		return FALSE;
+
+	/* wait for the device back */
+	if (!fu_device_retry_full(device,
+				  fu_mediatek_scaler_display_is_connected_cb,
+				  max_tries,
+				  FU_MEDIATEK_SCALER_DEVICE_POLL_INTERVAL,
+				  NULL,
+				  error)) {
+		g_prefix_error(error,
+			       "display controller did not reconnect after %u retries: ",
+			       max_tries);
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static FuFirmware *
+fu_mediatek_scaler_device_prepare_firmware(FuDevice *device,
+					   GBytes *fw,
+					   FwupdInstallFlags flags,
+					   GError **error)
+{
+	g_autoptr(FuFirmware) firmware = fu_firmware_new();
+	const guint8 *fw_ptr = g_bytes_get_data(fw, NULL);
+	const gsize fw_sz = g_bytes_get_size(fw);
+	guint32 ver_tmp = 0x0;
+	g_autofree gchar *dev_version = g_strdup(fu_device_get_version(device));
+	g_autofree gchar *fw_version = NULL;
+	g_autofree gchar *fw_date = NULL;
+	g_autofree gchar *fw_time = NULL;
+
+	/* read version from firmware */
+	if (!fu_memread_uint32_safe(fw_ptr,
+				    fw_sz,
+				    FIRMWARE_OFFSET_VERSION,
+				    &ver_tmp,
+				    G_LITTLE_ENDIAN,
+				    error))
+		return NULL;
+	fw_version = mediatek_scaler_device_version_to_string(ver_tmp);
+	g_info("firmware version: old: %s, new: %s", dev_version, fw_version);
+
+	/* read timestamp from firmware */
+	fw_date = fu_memstrsafe(fw_ptr,
+				fw_sz,
+				FIRMWARE_OFFSET_TIMESTAMP_DATE,
+				FIRMWARE_TIMESTAMP_DATE_SIZE,
+				error);
+	fw_time = fu_memstrsafe(fw_ptr,
+				fw_sz,
+				FIRMWARE_OFFSET_TIMESTAMP_TIME,
+				FIRMWARE_TIMESTAMP_TIME_SIZE,
+				error);
+	g_info("firmware timestamp: %s, %s", fw_time, fw_date);
```

Comment:
shouldn't this be in fu-mediatek-firmware.c?

Reply:
There isn't much to be parsed/tested in the firmware at this point so I put in a simple function, need a separate firmware.c?

Reply:
think it'll make the -device.c file much more manageable. It's also 100% to jut implement ->parse -- you don't need all the other stuff.